substrate/coins/pallet/src/lib.rs

```diff
@@ -0,0 +1,314 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[frame_support::pallet]
+pub mod pallet {
+  use frame_support::pallet_prelude::{*, ValueQuery, OptionQuery};
+  use frame_system::pallet_prelude::*;
+
+  use sp_runtime::{
+    traits::{DispatchInfoOf, PostDispatchInfoOf},
+    transaction_validity::{TransactionValidityError, InvalidTransaction},
+  };
+  use sp_std::vec::Vec;
+
+  use pallet_transaction_payment::{Config as TpConfig, OnChargeTransaction};
+
+  use serai_primitives::*;
+  pub use coins_primitives as primitives;
+  use primitives::*;
+
+  #[pallet::config]
+  pub trait Config: frame_system::Config<AccountId = PublicKey> + TpConfig {
+    type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+  }
+
+  #[pallet::genesis_config]
+  #[derive(Clone, PartialEq, Eq, Debug, Encode, Decode)]
+  pub struct GenesisConfig<T: Config> {
+    pub accounts: Vec<(T::AccountId, Coin, SubstrateAmount)>,
+  }
+
+  impl<T: Config> Default for GenesisConfig<T> {
+    fn default() -> Self {
+      GenesisConfig { accounts: Default::default() }
+    }
+  }
+
+  #[pallet::error]
+  pub enum Error<T> {
+    AmountOverflowed,
+    NotEnoughFunds,
+  }
+
+  #[pallet::event]
+  #[pallet::generate_deposit(fn deposit_event)]
+  pub enum Event<T: Config> {
+    Mint { address: SeraiAddress, balance: Balance },
+    Burn { address: SeraiAddress, balance: Balance, instruction: OutInstruction },
+    SriBurn { address: SeraiAddress, amount: Amount },
+    Transfer { from: SeraiAddress, to: SeraiAddress, balance: Balance },
+  }
+
+  #[pallet::pallet]
+  pub struct Pallet<T>(PhantomData<T>);
+
+  /// The amount of funds each account has.
+  #[pallet::storage]
+  #[pallet::getter(fn coins)]
+  pub type Coins<T: Config> = StorageDoubleMap<
+    _,
+    Blake2_128Concat,
+    T::AccountId,
+    Identity,
+    Coin,
+    SubstrateAmount,
+    OptionQuery,
+  >;
+
+  /// The total supply of each coin.
+  #[pallet::storage]
+  #[pallet::getter(fn supply)]
+  pub type Supply<T: Config> = StorageMap<_, Blake2_128Concat, Coin, SubstrateAmount, ValueQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+    fn build(&self) {
+      // initialize the supply of the coins
+      for c in COINS.iter() {
+        Supply::<T>::set(c, 0);
+      }
+
+      // initialize the genesis accounts
+      for (account, coin, amount) in self.accounts.iter() {
+        Coins::<T>::set(account, coin, Some(*amount));
+        let current_supply = Supply::<T>::get(coin);
+        Supply::<T>::set(coin, current_supply.checked_add(*amount).unwrap());
```

Comment:
These lines should be replaced with `mint`. They won't emit events right now *and* overwrite balance due to non-usage of `mint`.